diff -burpN -x cscope.out -x MakefileHost -x MakefileTarget ltp-full-20080630/testcases/kernel/sched/cfs-scheduler/hackbench.c ltp-full-20080630.modified/testcases/kernel/sched/cfs-scheduler/hackbench.c
--- ltp-full-20080630/testcases/kernel/sched/cfs-scheduler/hackbench.c	2008-06-28 15:19:08.000000000 -0500
+++ ltp-full-20080630.modified/testcases/kernel/sched/cfs-scheduler/hackbench.c	2008-07-25 10:47:08.000000000 -0500
@@ -48,6 +48,7 @@
 /*                                                                            */
 /******************************************************************************/
 #include <pthread.h>
+#include <limits.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
