--- ltp-20170929/runltp	2017-09-29 15:10:17.000000000 +0000
+++ ltp-20170929-patched/runltp	2017-12-21 20:28:01.870549897 +0000
@@ -172,7 +172,7 @@
     -z BIG_DEVICE   Some tests require a big unmounted block device
                     to run correctly.
     -Z  LTP_BIG_DEV_FS_TYPE The file system of the big device
-
+    -A ZOOFILE      Specify an alternate path to zoo file.


     example: ${0##*/} -c 2 -i 2 -m 2,4,10240,1 -D 2,10,10240,1 -p -q  -l /tmp/result-log.$$ -o /tmp/result-output.$$ -C /tmp/result-failed.$$ -d ${PWD}
@@ -219,11 +219,13 @@
     local PAN_COMMAND=
     local RANDOMRUN=0
     local DEFAULT_FILE_NAME_GENERATION_TIME=`date +"%Y_%m_%d-%Hh_%Mm_%Ss"`
+    # default value of zoofile is $$ to maintain normal functionality.
+    local ZOOFILE="$$"
     local scenfile=

     version_date=$(cat "$LTPROOT/Version")

-    while getopts a:b:B:c:C:T:d:D:ef:F:g:hi:I:K:l:m:M:No:pqQr:Rs:S:t:T:w:x:z:Z: arg
+    while getopts a:b:B:c:C:T:d:D:ef:F:g:hi:I:K:l:m:M:No:pqQr:Rs:S:t:T:w:x:z:Z:A: arg
     do  case $arg in
         a)  EMAIL_TO=$OPTARG
             ALT_EMAIL_OUT=1;;
@@ -447,6 +449,7 @@
         B) LTP_DEV_FS_TYPE=$OPTARG;;
         z) BIG_DEVICE=$OPTARG;;
         Z) BIG_DEVICE_FS_TYPE=$OPTARG;;
+        A)  ZOOFILE=$OPTARG;;
         \?) usage;;
         esac
     done
@@ -733,6 +736,8 @@
         fi
     fi

+    export ZOOFILE="$ZOOFILE"
+
     if [ $RUN_REPEATED -gt 1 ]; then # You need to specify at least more than 1 sequential run, else it runs default
          echo "PAN will run these test cases $RUN_REPEATED times....."
          echo "Test Tags will be Prepended with ITERATION NO.s....."
@@ -750,7 +755,7 @@
     fi

     [ ! -z "$QUIET_MODE" ] && { echo "INFO: Test start time: $(date)" ; }
-    PAN_COMMAND="${LTPROOT}/bin/ltp-pan $QUIET_MODE $NO_KMSG -e -S $INSTANCES $DURATION -a $$ \
+    PAN_COMMAND="${LTPROOT}/bin/ltp-pan $QUIET_MODE $NO_KMSG -e -S $INSTANCES $DURATION -a $ZOOFILE \
     -n $$ $PRETTY_PRT -f ${TMP}/alltests $LOGFILE $OUTPUTFILE $FAILCMDFILE $TCONFCMDFILE"
     echo "COMMAND:    $PAN_COMMAND"
     if [ ! -z "$TAG_RESTRICT_STRING" ] ; then
@@ -846,7 +851,7 @@
 	fi
     # Some tests need to run inside the "bin" directory.
     cd "${LTPROOT}/testcases/bin"
-    "${LTPROOT}/bin/ltp-pan" $QUIET_MODE $NO_KMSG -e -S $INSTANCES $DURATION -a $$ -n $$ $PRETTY_PRT -f ${TMP}/alltests $LOGFILE $OUTPUTFILE $FAILCMDFILE $TCONFCMDFILE
+    "${LTPROOT}/bin/ltp-pan" $QUIET_MODE $NO_KMSG -e -S $INSTANCES $DURATION -a $ZOOFILE -n $$ $PRETTY_PRT -f ${TMP}/alltests $LOGFILE $OUTPUTFILE $FAILCMDFILE $TCONFCMDFILE

     if [ $? -eq 0 ]; then
       echo "INFO: ltp-pan reported all tests PASS"
